### ISSUE TITLE
File(path.m4a) broken on linux

### ISSUE THREAD

**Issue by dreamflasher** [quodlibet/mutagen#659]
My code:

```
File("audio.mp3")
File("audio.m4a")
```

Observation:
Works as expected on Windows, meaning: I get duration and tags for both the mp3 and m4a file. 
On Linux only `File("audio.mp3")` works. `File("audio.m4a")` is None. The mp3 and the m4a are different (they have the correct format and correct file ending), but the m4a is identical on both OS (the mp3 as well).

Unfortunately, there is no error message indicating what the issue could be.

**Comment by phw** [comment]
Which versions of mutagen are you using? Is it the same version on both environments?

**Comment by dreamflasher** [comment]
In freshly installed mutagen on both OS yesterday, so I expect it to be the latest. Unfortunately mutagen does not populate `mutagen.__version__` – how do I find out the version of the mutagen lib that's loaded?

I can reproduce the issue with `mutagen-inspect`:
Windows output:
```
- MPEG-4 audio (AAC LC), 3914.72 seconds, 126250 bps (audio/mp4)
©too=Lavf61.1.100
```

Linux:
```
- Unknown file type
```

**Comment by phw** [comment]
Check `mutagen.version`.

> I freshly installed mutagen on both OS yesterday, so I expect it to be the latest. 

It depends how and from where you installed mutagen. If you installed it from the package repository of your Linux distribution you might well have an older version, depending on the distribution you use.

**Comment by dreamflasher** [comment]
```
>>> mutagen.version
(1, 47, 0)
```

on both OS, installed via pip

**Comment by phw** [comment]
This is rather strange, I don't know why there is such a difference between OS versions.

Are you sure the file is actually a MP4 file? The test with file extension changed to `.aac`  indicates this might actually be a raw AAC file.

Can you try opening the file with both `f = mutagen.mp4.MP4('test.m4a')` and `f = mutagen.aac.AAC('test.m4a')`, and should it succeed look at the output of `f.info.pprint()`?

> I tried renaming the file extension, I'm surprised that this results in different outputs (would have expected that the file extension is ignored):

Depending on format the scoring functions used to determine the file type might take the file extension into account, in addition of looking at format typical bytes at the start of the file.

**Comment by dreamflasher** [comment]
I run this code on both OS, with the identic file, just renamed extension:

```
import mutagen
import mutagen.aac
import mutagen.mp4

for f in ["audio.mp4", "audio.m4a", "audio.aac"]:
    for t in [mutagen.File, mutagen.mp4.MP4, mutagen.aac.AAC]:
        try:
            print("\n", f, t, t(f))
            t(f).info.pprint()
        except Exception as e:
            print("\n", f, t, e)
```

Ouput on Windows:
```
 audio.mp4 <function File at 0x00000276F8B7FF60> None

 audio.mp4 <function File at 0x00000276F8B7FF60> 'NoneType' object has no attribute 'info'

 audio.mp4 <class 'mutagen.mp4.MP4'> not a MP4 file

 audio.mp4 <class 'mutagen.aac.AAC'> {}
'AAC (ADTS), 32000 Hz, 3908.46 seconds, 1 channel(s), 126455 bps'

 audio.m4a <function File at 0x00000276F8B7FF60> None

 audio.m4a <function File at 0x00000276F8B7FF60> 'NoneType' object has no attribute 'info'

 audio.m4a <class 'mutagen.mp4.MP4'> not a MP4 file

 audio.m4a <class 'mutagen.aac.AAC'> {}
'AAC (ADTS), 32000 Hz, 3908.46 seconds, 1 channel(s), 126455 bps'

 audio.aac <function File at 0x00000276F8B7FF60> {}
'AAC (ADTS), 32000 Hz, 3908.46 seconds, 1 channel(s), 126455 bps'

 audio.aac <class 'mutagen.mp4.MP4'> not a MP4 file

 audio.aac <class 'mutagen.aac.AAC'> {}
'AAC (ADTS), 32000 Hz, 3908.46 seconds, 1 channel(s), 126455 bps'
```

When I run it on Ubuntu I get the same, so by now I can't tell what I did that made me believe that the results are different on both OS. 
This renders my original issue title incorrect, but still the library behaves unexpected to me: I believe the file extension should not matter, if it's named mp4 and one of the codecs can read the file it should do so.

**Comment by phw** [comment]
Thanks for the details, this resolves the confusion. So this is indeed a raw AAC file with a misleading file extension (.m4a indicates it is an MP4 container), which the automatic file type detection fails to identify.

The AAC implementation currently only has insufficient file type identification and primarily relies on file extensions to be .aac, .adts or .adif. This is a known issue which should be fixed, see #394 . If this gets addressed the detection should work. What would be interesting to see are the first few bytes from your file.

Just because this might come up as the next question: Raw AAC files do not directly support tags, and mutagen will raise an error if you try to call `add_tags()` on the file object.

Some taggers and audio players support tucking APE or ID3 tags at the end of the file. This however can cause problems with players not expecting it. E.g. some players will wrongly calculate the track length and trigger errors when trying to playback the data from the tag block, or they might completely fail to load and play the file. If you want to see some implementation of using AAC files with appended APEv2 tags using mutagen see https://github.com/metabrainz/picard/blob/master/picard/formats/mutagenext/aac.py

The much better solution however is to wrap the AAC into a MP4 container, e.g. using ffmpeg. There is some discussion about this on https://community.mp3tag.de/t/aac-tagging-corrupts-the-music/53597

**Comment by dreamflasher** [comment]
Thank you very much for the detailed answer @phw . I will close this issue as a duplicate of #394 . 
The first bytes are:
`b'\xff\xf1T@@\x1f\xfc\x01@"\x80\xa3\x7f\xf8\x85'`

Shall I also post this in the other issue?
